Improve docs for custom events

jrieke opened the pull request:
Fixes #1555 

Description: Improved [the docs](https://pytorch.org/ignite/concepts.html#events-and-handlers) on how to create custom events according to the example in #1555.


Check list:
* [ ] New tests are added (if a new feature is added)
* [x] New doc strings: description and/or example code are in RST format
* [x] Documentation is updated (if required)

vfdev-5 commented on docs/source/concepts.rst:
```diff
@@ -254,6 +260,7 @@ These events could be used to attach any handler and are fired using :meth:`~ign
     @engine.on(Events.STARTED)
     def fire_custom_events(engine):
          engine.fire_event(CustomEvents.CUSTOM_STARTED)
+         engine.state.custom_started += 1  # increase state counter for this event (required)
```
Let's say this is required for filtering only and can be omitted if no filtering is done. And another note to add is that this is a temporary solution and subject to change in future releases.

vfdev-5 commented on docs/source/concepts.rst:
```diff
@@ -241,9 +241,15 @@ and be registered with :meth:`~ignite.engine.engine.Engine.register_events` in a
-    engine.register_events(*CustomEvents)
+    engine.register_events(
+        *CustomEvents,
+        event_to_attr={
```
This part is not required actually. I'd put everything in the following way:
- simple custom events without filtering (current doc)
- custom events with filtering:
   - engine.state should have corresponding attributes
   - event_to_attr is required when `register_events`
   - should increase counters (subject to be removed)

vfdev-5 reviewed (commented):
Thanks for the PR @jrieke !
I left few comments to improve things

jrieke commented:
@vfdev-5 Ok I restructured it as you suggested. I removed all changes from the original docs and added the 3 steps for filtering as a paragraph below it.

vfdev-5 commented on docs/source/concepts.rst:
```diff
@@ -260,6 +260,25 @@ These events could be used to attach any handler and are fired using :meth:`~ign
+
+- you should increase the counter for the event whenever you fire the event, e.g. `engine.state.custom_started += 1`
+
+.. Note ::
```
```suggestion
.. warning::
```

vfdev-5 reviewed (commented):
A minor fix to add, otherwise it is OK. Thanks @jrieke !

vfdev-5 commented on docs/source/concepts.rst:
```diff
@@ -260,6 +260,25 @@ These events could be used to attach any handler and are fired using :meth:`~ign
+
+.. code-block:: python
+
+    event_to_attr={
+            CustomEvents.CUSTOM_STARTED: "custom_started",
+            CustomEvents.CUSTOM_COMPLETED: "custom_completed",
+        }
```
```suggestion
    event_to_attr = {
        CustomEvents.CUSTOM_STARTED: "custom_started",
        CustomEvents.CUSTOM_COMPLETED: "custom_completed",
    }
```

vfdev-5 commented on docs/source/concepts.rst:
```diff
@@ -260,6 +260,25 @@ These events could be used to attach any handler and are fired using :meth:`~ign
 
+If you want to use filtering with custom events (e.g. with `CallableEventWithFilter`), you need to do 3 more things:
+
+- `engine.state` should have corresponding attributes for the events, e.g. `engine.state.custom_started`
```
```suggestion
- ``engine.state`` should have corresponding attributes for the events, e.g. ``engine.state.custom_started``
```

vfdev-5 commented on docs/source/concepts.rst:
```diff
@@ -260,6 +260,25 @@ These events could be used to attach any handler and are fired using :meth:`~ign
+If you want to use filtering with custom events (e.g. with `CallableEventWithFilter`), you need to do 3 more things:
+
+- `engine.state` should have corresponding attributes for the events, e.g. `engine.state.custom_started`
+- you need to pass a dict `event_to_attr` to `register_events`, which maps between events and state attributes, e.g.
```
```suggestion
- you need to pass a dict `event_to_attr` to :meth:`~ignite.engine.engine.Engine.register_events`, which maps between events and state attributes, e.g.
```

vfdev-5 commented on docs/source/concepts.rst:
```diff
@@ -260,6 +260,25 @@ These events could be used to attach any handler and are fired using :meth:`~ign
+            CustomEvents.CUSTOM_COMPLETED: "custom_completed",
+        }
+
+- you should increase the counter for the event whenever you fire the event, e.g. `engine.state.custom_started += 1`
```
```suggestion
- you should increase the counter for the event whenever you fire the event, e.g. ``engine.state.custom_started += 1``
```

vfdev-5 commented on docs/source/concepts.rst:
```diff
@@ -260,6 +260,25 @@ These events could be used to attach any handler and are fired using :meth:`~ign
    See the source code of :class:`~ignite.contrib.engines.create_supervised_tbptt_trainer` for an example of usage of
    custom events.
 
+If you want to use filtering with custom events (e.g. with `CallableEventWithFilter`), you need to do 3 more things:
```
```suggestion
If you want to use filtering with custom events (e.g. with ``CustomEvents.CUSTOM_STARTED(every=5)``), you need to do 3 more things:
```

vfdev-5 commented on docs/source/concepts.rst:
```diff
@@ -260,6 +260,25 @@ These events could be used to attach any handler and are fired using :meth:`~ign
    See the source code of :class:`~ignite.contrib.engines.create_supervised_tbptt_trainer` for an example of usage of
    custom events.
 
+If you want to use filtering with custom events (e.g. with `CallableEventWithFilter`), you need to do 3 more things:
```
Can you also please add a subsection to the line 229 : 

    Custom events
    ``````````````

jrieke replied:
Done!

vfdev-5 reviewed (commented):
Few more comments

vfdev-5 commented on docs/source/concepts.rst:
```diff
@@ -260,6 +263,25 @@ These events could be used to attach any handler and are fired using :meth:`~ign
    See the source code of :class:`~ignite.contrib.engines.create_supervised_tbptt_trainer` for an example of usage of
    custom events.
 
+If you want to use filtering with custom events (e.g. with ``CustomEvents.CUSTOM_STARTED(every=5)``), you need to do 3 more things:
```
```suggestion
If you want to use filtering with custom events (e.g. ``CustomEvents.CUSTOM_STARTED(every=5)``), you need to do 3 more things:
```

vfdev-5 reviewed (commented):
LGTM, thanks @jrieke !
